### PR TITLE
Bipartite reader writer

### PR DESCRIPTION
**Pull request by hagberg**
Addresses #1294.

### CODE REVIEW

**Comment by jtorrents** [comment]
Wow. I missed this. It's really great @hagberg ! I think that the only thing that should be changed is the name of the node attribute that indicates the node set. It's `part` and should be `bipartite` for consistency.

File: networkx/algorithms/bipartite/edgelist.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    2 3 {'capacity': 12}
+
+    >>> for line in bipartite.generate_edgelist(G,data=['weight']):
+    ...     print(line)
+    0 1
+    2 1 3
+    2 3
+    """
+    try:
+        part0 = [n for n,d in G.node.items() if d['part'] == 0]
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/edgelist.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/edgelist.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    -------
+    G : graph
+       A networkx Graph or other type specified with create_using
+
+    Examples
+    --------
+    >>> from networkx.algorithms import bipartite
+    >>> G = nx.path_graph(4)
+    >>> G.add_nodes_from([0,2], part=0)
+    >>> G.add_nodes_from([1,3], part=1)
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/edgelist.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/matrix.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+    References
+    ----------
+    [1] http://en.wikipedia.org/wiki/Adjacency_matrix#Adjacency_matrix_of_a_bipartite_graph
+    """
+    G = _prep_create_using(create_using)
+    n, m = A.shape
+    # Make sure we get even the isolated nodes of the graph.
+    G.add_nodes_from(range(n), part=0)
+    G.add_nodes_from(range(n,n+m), part=1)
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/matrix.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/tests/test_edgelist.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+from networkx.algorithms import bipartite
+
+class TestEdgelist:
+
+    def setUp(self):
+        self.G=nx.Graph(name="test")
+        e=[('a','b'),('b','c'),('c','d'),('d','e'),('e','f'),('a','f')]
+        self.G.add_edges_from(e)
+        self.G.add_nodes_from(['a','c','e'],part=0)
+        self.G.add_nodes_from(['b','d','f'],part=1)
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/tests/test_edgelist.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/tests/test_edgelist.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        assert_equal(fh.read(),b"1 2 {}\n3 2 {}\n")
+
+    def test_write_edgelist_3(self):
+        fh=io.BytesIO()
+        G=nx.Graph()
+        G.add_edge(1,2,weight=2.0)
+        G.add_edge(2,3,weight=3.0)
+        G.add_node(1,part=0)
+        G.add_node(2,part=1)
+        G.add_node(3,part=0)
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/tests/test_edgelist.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/tests/test_edgelist.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        assert_equal(fh.read(),b"1 2 {'weight': 2.0}\n3 2 {'weight': 3.0}\n")
+
+    def test_write_edgelist_4(self):
+        fh=io.BytesIO()
+        G=nx.Graph()
+        G.add_edge(1,2,weight=2.0)
+        G.add_edge(2,3,weight=3.0)
+        G.add_node(1,part=0)
+        G.add_node(2,part=1)
+        G.add_node(3,part=0)
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/tests/test_edgelist.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/tests/test_edgelist.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        G = nx.Graph()
+        try: # Python 3.x
+            name1 = chr(2344) + chr(123) + chr(6543)
+            name2 = chr(5543) + chr(1543) + chr(324)
+        except ValueError: # Python 2.6+
+            name1 = unichr(2344) + unichr(123) + unichr(6543)
+            name2 = unichr(5543) + unichr(1543) + unichr(324)
+        G.add_edge(name1, 'Radiohead', attr_dict={name2: 3})
+        G.add_node(name1,part=0)
+        G.add_node('Radiohead',part=1)
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/tests/test_edgelist.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/tests/test_edgelist.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/tests/test_edgelist.py]: Node attribute should be `bipartite`.

File: networkx/algorithms/bipartite/tests/test_edgelist.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        try: # Python 3.x
+            blurb = chr(1245) # just to trigger the exception
+            name1 = 'Bj' + chr(246) + 'rk'
+            name2 = chr(220) + 'ber'
+        except ValueError: # Python 2.6+
+            name1 = 'Bj' + unichr(246) + 'rk'
+            name2 = unichr(220) + 'ber'
+        G.add_edge(name1, 'Radiohead', attr_dict={name2: 3})
+        G.add_node(name1,part=0)
+        G.add_node('Radiohead',part=1)
```
Comment (jtorrents) [inline networkx/algorithms/bipartite/tests/test_edgelist.py]: Node attribute should be `bipartite`.

**Comment by hagberg** [comment]
I think I got all of the part->bipartite fixes.

**Comment by jtorrents** [comment]
Looks good to me. I think is ready to merge.

**Comment by chebee7i** [comment]
Did this get merged?

**Comment by chebee7i** [comment]
Yep: Merged in f634c4f6de8db8ab42f191af09bb328a6ec78623

**Comment by hagberg** [comment]
Also broke the build.  Should have just fixed that in https://github.com/networkx/networkx/commit/d15361da366b77573777a1542741387c29f34717
Need more coffee.

**Pull request closed without merging**